### DISCUSSION TITLE
Can't check out a different branch on a submodule

### DISCUSSION THREAD

**Discussion by aormsby** [community/community#26818] [Actions]
<p>I’m using <code>actions/checkout@v2</code> to check out a repo and all its submodules. or From my workflow:</p>
<pre><code class="lang-yaml">- name: Checkout repo with submodules (theme and public)
      uses: actions/checkout@v2
      with:
        token: ${{ secrets.SOME_PAT }}
        submodules: recursive
</code></pre>
<p>The main repo checks out the <code>master</code> branch, but my submodule checks out the latest commit as a detached HEAD. The submodule <code>master</code> also shows up when I list branches, but I have to explicitly check it out if I want that.</p>
<p>However, <strong>none of the other branches appear</strong> for the submodule, and <strong>I can’t find a way to check them out</strong> when I need to. I don’t see that I can do this through the checkout Action, but I’ve tried some other things I’d expect to at least do <em>something</em>. In the next Action step, I’ve tried these commands on the submodule.</p>
<pre><code class="lang-shell">git -C subfolder fetch --all
git -C subfolder remote -v update
git -C subfolder pull --all
</code></pre>
<p>But no matter what I do, I only see these:</p>
<pre><code class="lang-auto">$ git branch

* (HEAD detached at 01e4b10)
  master
</code></pre>
<p><strong>None</strong> of my other branches ever show up, and I can’t check them out. Any idea why? Is there some limitation in Actions with submodules? Am I just doing something silly and wrong?</p>

**Comment by aormsby** [comment]
<p>Okay, I’ve tried lots of new things, and I figured <em>something</em> out. I can see the remote refs like this:</p>
<pre><code class="lang-shell">$ git ls-remote refs
[some hash number]        refs/heads/master
[some hash number]        refs/heads/test1
</code></pre>
<p>So the other branch does exist remotely. But still a branch list doesn’t show <code>test1</code>, so I can’t switch to it or fetch it.</p>
<pre><code class="lang-shell">$ git branch -a
* master
  remotes/origin/HEAD -&gt; origin/master
  remotes/origin/master
</code></pre>
<p>This <em><strong>only</strong></em> occurs after using the checkout Action. When I clone with submodules locally everything is fine. I just don’t know if this is a bug for the checkout action or if I’m doing something wrong.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/aormsby">@aormsby</a>,</p>
<p>Submodules are not on a branch. They are just a pointer to a particular commit of the submodule’s repository.</p>
<p>After checkout the parent repo, you can update the submodule branch, and checkout the target branch then. Code sample as below(submodule : ticket7 with B1 branch)</p>
<pre><code class="lang-auto">      - name: checkout
        uses: actions/checkout@v2
        with:
          token: ${{ secrets.GITHUB_TOKEN }}    # you can use PAT token here.
      - name: update submodule branch
        run: |
           git config --file=.gitmodules submodule.Submod.branch B1  # change branch to 'B1' branch
           git submodule sync                        # sync
           git submodule update --init --recursive --remote   # pull the submodule
      - name: checkout submodule B1 branch
        run: |
          cd ticket7    # go to submodule repo
          git checkout -b B1 origin/B1   # checkout the target branch.
</code></pre>
<p>Please refer to my <a href="https://github.com/weide-zhou/ticket10/runs/1011358671?check_suite_focus=true" rel="nofollow noopener">workflow</a> for the details.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/1/173929b61b070095df54ae3656c382f8ecdb7f36.png" data-download-href="/uploads/short-url/3jrs4JjWnsRqVvtqoEjUvcZ4D2K.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181100026-f23ae4c7-a718-4084-9497-0ff1cd413e29.png" alt="image" data-base62-sha1="3jrs4JjWnsRqVvtqoEjUvcZ4D2K" width="690" height="269" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/1/173929b61b070095df54ae3656c382f8ecdb7f36_2_690x269.png, https://aws1.discourse-cdn.com/github/optimized/2X/1/173929b61b070095df54ae3656c382f8ecdb7f36_2_1035x403.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/1/173929b61b070095df54ae3656c382f8ecdb7f36.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/1/173929b61b070095df54ae3656c382f8ecdb7f36_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1321×516 31.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
Doc: <a href="https://tech.serhatteker.com/post/2019-01/changing-git-submodules-urlbranch-to/" rel="nofollow noopener">https://tech.serhatteker.com/post/2019-01/changing-git-submodules-urlbranch-to/</a>
<p>Thanks</p>

**Comment by aormsby** [comment]
<p>Ooooohhhhhh, this makes complete sense. I just didn’t realize I needed to set the <code>branch</code> property of .gitmodules to sync it that way. Nice solution.</p>
<p>There’s still one small snag for me. If the submodule is a private repo, I get a fatal error when I run the sync because the submodule URL doesn’t include user name and key information by default.</p>
<pre><code class="lang-auto">Cloning into '/home/runner/work/my/submod/path'...
fatal: could not read Username for 'https://github.com': No such device or address
fatal: clone of 'https://github.com/user/repo.git' into submodule path '/home/runner/work/my/submod/path' failed
</code></pre>
<p>I <em>may</em> be able to manipulate the URL in .gitmodules to include that, but I think it would be better to use the secret token like in the checkout. Shouldn’t the token credentials persist from the checkout action?? The readme says they should by default, but maybe I’m doing something wrong there as well.</p>
<p>Edit: Also, could you explain why keeping <code>submodules: recursive</code> in the workflow causes a fatal error? The update command results in <code>fatal: Needed a single revision</code></p>

**Comment by aormsby** [comment, marked as answer]
<p>Okay, I’ve got it! The original solution I marked got me really close to the right functionality, but there were still some issues coming through. However, I want to thank <a class="mention" href="/weide-zhou">@weide-zhou</a> for their help. I couldn’t have come to these solutions without the push. :slight_smile:</p>
<h3>The Underlying Problem</h3>
<p>The <a href="https://git-scm.com/book/en/v2/Git-Internals-The-Refspec" rel="nofollow noopener">git refspec</a> for the submodules wasn’t configured to track any remote refs outside ‘master’ so I could never get the other remote refs to other branches.</p>
<p><strong>Incorrect config:</strong></p>
<pre><code class="lang-nohighlight">[remote "origin"]
	fetch = +refs/heads/master:refs/remotes/origin/master
# tied to master refs only
</code></pre>
<p><strong>Config that gets all refs:</strong></p>
<pre><code class="lang-nohighlight">[remote "origin"]
	fetch = +refs/heads/*:refs/remotes/origin/*
# gets all refs - *
</code></pre>
<p>The checkout action has a default <code>fetch-depth: 1</code>, which avoids pulling all the history of the repo. However, setting a specific depth triggers some command options during the checkout that limits all fetch refs to the ‘master’ branch in the config (or whatever other ref you set in the workflow).</p>
<p>So we gotta fix that. <a class="mention" href="/weide-zhou">@weide-zhou</a> was on the way to that, but I found two ways to do it that attack the problem more directly.</p>
<h3>Solution <span class="hashtag">#1</span> - change fetch-depth</h3>
<p>Workflow:</p>
<pre><code class="lang-yaml">steps:
    - name: Checkout repo with submodules (theme and public)
      uses: actions/checkout@v2
      with:
        token: ${{ secrets.YOUR_PAT}}  # for private repos (including submodules)
        submodules: 'recursive'
        fetch-depth: '0'    ***********
</code></pre>
<p>From the checkout readme:</p>
<blockquote>
<p>0 indicates all history for all branches and tags.</p>
</blockquote>
<p>Setting the value to 0 leaves the fetch config set to *, and all remote refs are available for checkout. <strong>This is the quick fix.</strong> The downside is that you’re pulling the entire history of your branches and tags for the main repo and all submodules, so I also came up with a more surgical adjustment.</p>
<h3>Solution <span class="hashtag">#2</span> - modify the config fetch by hand</h3>
<p>This solution allows control over specific submodule fetch config so other submodule configurations are left untouched.</p>
<p>Workflow (same minus fetch-deptch):</p>
<pre><code class="lang-yaml">steps:
    - name: Checkout repo with submodules (theme and public)
      uses: actions/checkout@v2
      with:
        token: ${{ secrets.YOUR_PAT}}  # for private repos (including submodules)
        submodules: 'recursive' 
</code></pre>
<p>Somewhere in your action you can run the following commands:</p>
<pre><code class="lang-shell">$ git -C submodPath config remote.origin.fetch +refs/heads/*:refs/remotes/origin/*
## yeah, that's it!

## after that key config update, the rest is what @weide-zhou recommended
$ git submodule sync --recursive submodPath
$ git submodule update --init --recursive --remote submodPath
$ git -C public checkout myBranch
</code></pre>
<ul>
<li>
<code>submodPath</code> should be replaced with the path of the submodule from the root</li>
<li>The * will allow fetching all refs in the submodule, but you can also replace it with the specific branch name you want if that’s all you need.</li>
</ul>
<p>The main reason I had to build on <a class="mention" href="/weide-zhou">@weide-zhou</a>’s solution was to maintain the <code>submodules: 'recursive'</code> setting in the workflow. It’s easier to manage general submodule checkouts in the first step than to have to deal with them in my following actions. But yeah, mostly the same! I just needed that key change in the ref config.</p>

**Comment by lermana** [comment]
<p><a class="mention" href="/aormsby">@aormsby</a> thank you so much! your back-and-forth (plus the detailed examples) saved me a whole bunch of time.</p>
<p>I wanted to add a couple of notes in case anyone else finds themselves here:</p>
<ol>
<li>in <a class="mention" href="/aormsby">@aormsby</a>’s last comment, note that <code>-C</code> means that <code>git</code> is executing as if it was within the path that comes after it; i.e. commands are running from within the submodule without having to <code>cd</code> back and forth</li>
<li>the command <code>git -C public checkout myBranch</code> should (I think) have been written as <code>git -C submodPath checkout myBranch</code>
</li>
</ol>
<p>Here is an excerpt from my workflow, which checks out a submodule parent, updates the reference to a particular submodule, and pushes that out, in case anyone finds this helpful. My goal here is to have a workflow actually running from within the submodule / child repository that checks out its referencing parent and programmatically updates the reference its parent keeps to whatever the current branch is:</p>
<pre><code class="lang-auto">steps:

      - name: Checkout parent
        uses: actions/checkout@v2
        with:
          repository: owner/repo # these are fake -- must update
          token: ${{ secrets.GITHUB_TOKEN }}
          submodules: true # we in this case only want to update the specific submodule, not all
          path: parent

      - name: Update parent
        working-directory: parent
        run: | 
          echo "Updating parent"

          git config user.name git_user # fake, update
          git config user.email git_email # fake, update
          echo "git configuration complete"

          branch=$( echo ${{ github.ref }} | sed 's/refs\/heads\///g' )
          echo "working on parent branch ${branch}"
          if git rev-parse --quiet --verify $branch &gt; /dev/null; then git checkout $branch; else git checkout -b $branch; fi;

          package=$(echo $GITHUB_REPOSITORY | sed 's:.*/::')
          echo "updating sudmodule for ${package}"

          git -C $package config remote.origin.fetch +refs/heads/*:refs/remotes/origin/*
          git submodule update --remote $package

          git -C $package checkout $branch
          echo "working on package branch ${branch}"

          git add .
          git commit -m "update reference to $package"
          git fetch
          git branch --set-upstream-to=origin/$branch
          
          git pull --no-edit
          git push
</code></pre>

**Comment by aormsby** [comment]
<p>Turns out you can also do this:</p>
<pre><code class="lang-shell"># fetch the specific ref
git -C &lt;submod-directory&gt; fetch --depth=1 origin "refs/heads/&lt;branch&gt;"

# do anything with it now!
git checkout &lt;branch&gt;
</code></pre>
<p>No idea what’s better/faster/stronger here, but this solution was definitely straightforward. Works for any ref you don’t already have access to from the initial checkout. :slight_smile:</p>
<p><strong>Edit:</strong> I do want to be clear, though. Submodules are a pain, and different approaches work for different situations. For instance, my example here breaks my current work, and it’s annoying. Just something to be aware of.</p>
